File: capstone-assignment/src/location.cpp
#include "location.h"

std::string Location::to_string(Location::Country country)
{
    switch (country)
    {
        case Country::UnitedStates:
            return "UnitedStates";
        case Country::Canada:
            return "Canada";
        case Country::UnitedKingdom:
            return "UnitedKingdom";
        case Country::Germany:
            return "Germany";
        case Country::France:
            return "France";
        case Country::Japan:
            return "Japan";
        case Country::Australia:
            return "Australia";
        case Country::China:
            return "China";
        case Country::Brazil:
            return "Brazil";
        case Country::SouthKorea:
            return "South Korea";
        case Country::Ireland:
            return "Ireland";
        case Country::Spain:
            return "Spain";
        case Country::India:
            return "India";
        case Country::Switzerland:
            return "Switzerland";
        default:
            return "UnknownCountry";
    }
}

std::string Location::to_string(Location::Continent continent)
{
    switch (continent)
    {
        case Continent::NorthAmerica:
            return "North America";
        case Continent::SouthAmerica:
            return "South America";
        case Continent::Europe:
            return "Europe";
        case Continent::Asia:
            return "Asia";
        case Continent::Africa:
            return "Africa";
        case Continent::Oceania:
            return "Oceania";
        default:
            return "UnknownContinent";
    }
}

Location::Country Location::from_string(std::string_view str)
{
    if (str == "UnitedStates" || str == "United States" || str == "USA")
        return Country::UnitedStates;
    else if (str == "Canada")
        return Country::Canada;
    else if (str == "UnitedKingdom" || str == "United Kingdom" || str == "UK")
        return Country::UnitedKingdom;
    else if (str == "Germany")
        return Country::Germany;
    else if (str == "France")
        return Country::France;
    else if (str == "Japan")
        return Country::Japan;
    else if (str == "Australia")
        return Country::Australia;
    else if (str == "China")
        return Country::China;
    else if (str == "Brazil")
        return Country::Brazil;
    else if (str == "SouthKorea" || str == "South Korea")
        return Country::SouthKorea;
    else if (str == "Ireland")
        return Country::Ireland;
    else if (str == "Spain")
        return Country::Spain;
    else if (str == "India")
        return Country::India;
    else if (str == "Switzerland")
        return Country::Switzerland;
    else
        return Country::UnknownCountry;
}

Location::Continent Location::country_to_continent(Location::Country country)
{
    switch (country)
    {
        case Country::UnitedStates:
        case Country::Canada:
            return Continent::NorthAmerica;
        case Country::UnitedKingdom:
        case Country::Germany:
        case Country::France:
        case Country::Ireland:
        case Country::Spain:
        case Country::Switzerland:
            return Continent::Europe;
        case Country::Japan:
        case Country::China:
        case Country::SouthKorea:
        case Country::India:
            return Continent::Asia;
        case Country::Australia:
            return Continent::Oceania;
        case Country::Brazil:
            return Continent::SouthAmerica;
        default:
            return Continent::UnknownContinent;
    }
}

std::string Location::continent_to_string(Location::Continent continent)
{
    switch (continent)
    {
        case Continent::NorthAmerica:
            return "NorthAmerica";
        case Continent::SouthAmerica:
            return "SouthAmerica";
        case Continent::Europe:
            return "Europe";
        case Continent::Asia:
            return "Asia";
        case Continent::Africa:
            return "Africa";
        case Continent::Oceania:
            return "Oceania";
        default:
            return "Unknown";
    }
}

File: capstone-assignment/src/location.rs


#[derive(Debug, PartialEq)]
//enum for country
pub enum Country
{
    UnitedStates,
    Canada,
    UnitedKingdom,
    Germany,
    France,
    Japan,
    Australia,
    China,
    Brazil,
    SouthKorea,
    Ireland,
    Spain,
    India,
    Switzerland
}


#[derive(Debug, PartialEq)]
//enum for continent
pub enum Continent
{
    NorthAmerica,
    SouthAmerica,
    Europe,
    Asia,
    Africa,
    Oceania
}

impl Continent
{
    //converts continent to string
    pub fn to_string(&self) -> String
    {
        match self
        {
            Continent::NorthAmerica => "North America".to_string(),
            Continent::SouthAmerica => "South America".to_string(),
            Continent::Europe => "Europe".to_string(),
            Continent::Asia => "Asia".to_string(),
            Continent::Africa => "Africa".to_string(),
            Continent::Oceania => "Oceania".to_string()
        }
    }
}

impl Country
{
    //converts country to continent
    pub fn country_to_continent(&self) -> Continent{
        match self{
            Country::UnitedStates => Continent::NorthAmerica,
            Country::Canada => Continent::NorthAmerica,
            Country::UnitedKingdom => Continent::Europe,
            Country::Germany => Continent::Europe,
            Country::France => Continent::Europe,
            Country::Japan => Continent::Asia,
            Country::Australia => Continent::Oceania,
            Country::China => Continent::Asia,
            Country::Brazil => Continent::SouthAmerica,
            Country::SouthKorea => Continent::Asia,
            Country::Ireland => Continent::Europe,
            Country::Spain => Continent::Europe,
            Country::India => Continent::Asia,
            Country::Switzerland => Continent::Europe
        }
    }
    //converts country to string
    pub fn to_string(&self) -> String
    {
        match self
        {
            Country::UnitedStates => "United States".to_string(),
            Country::Canada => "Canada".to_string(),
            Country::UnitedKingdom => "United Kingdom".to_string(),
            Country::Germany => "Germany".to_string(),
            Country::France => "France".to_string(),
            Country::Japan => "Japan".to_string(),
            Country::Australia => "Australia".to_string(),
            Country::China => "China".to_string(),
            Country::Brazil => "Brazil".to_string(),
            Country::SouthKorea => "South Korea".to_string(),
            Country::Ireland => "Ireland".to_string(),
            Country::Spain => "Spain".to_string(),
            Country::India => "India".to_string(),
            Country::Switzerland => "Switzerland".to_string()
        }
    }
}

//converts string to country
impl std::str::FromStr for Country{
    type Err = &'static str;
    fn from_str(s: &str) -> Result<Self, Self::Err>
    {
        match s
        {
            "USA" => Ok(Country::UnitedStates),
            "Canada" => Ok(Country::Canada),
            "UK" => Ok(Country::UnitedKingdom),
            "Germany" => Ok(Country::Germany),
            "France" => Ok(Country::France),
            "Japan" => Ok(Country::Japan),
            "Australia" => Ok(Country::Australia),
            "China" => Ok(Country::China),
            "Brazil" => Ok(Country::Brazil),
            "South Korea" => Ok(Country::SouthKorea),
            "Ireland" => Ok(Country::Ireland),
            "Spain" => Ok(Country::Spain),
            "India" => Ok(Country::India),
            "Switzerland" => Ok(Country::Switzerland),
            _ => Err("Invalid country name!")
        }
    }
}